### ISSUE TITLE
Record constructor arguments should propagate nullability to properties

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#44763]
**Version Used**:
f36e10a

**Steps to Reproduce**:
```cs
#nullable enable
data class B(string? X, string? Y)
{
    static void M1(bool flag)
    {
        B b = new B(""hello"", null);
        if (flag)
        {
            b.X.ToString(); // shouldn't warn
            b.Y.ToString(); // 1
        }

        b = b with { Y = ""world"" };
        b.X.ToString(); // shouldn't warn
        b.Y.ToString();
    }
}
``` 

**Expected Behavior**:
Warning only on line `// 1`

**Actual Behavior**:
Warnings on line `// 1` as well as `// shouldn't warn` lines

**Comment by RikkiGibson** [comment]
The main blocker on this is that we don't have a decent way to tell if a constructor is the record primary constructor. We can in source, but not so easily in metadata. We could go forward by just assuming that every record constructor parameter assigns to an identically-named property, but this seems dubious.

**Comment by RikkiGibson** [comment]
Solution to this issue should also address deconstruction as described in #49741

**Comment by RikkiGibson** [comment]
Closing as stale. It looks like we haven't had any requests to actually do this since #49741.
